JDK-8299229: [JVMCI] add support for UseZGC

tkrodriguez opened the pull request:
This exposes the required ZGC values to JVMCI and adds support for nmethod entry barriers.  The ZGC support is straightforward but the nmethod entry barrier required some reworking to fit better into JVMCI usage.  I also removed the epoch based barrier since it was no longer used with simplified the assumptions on the JVMCI side.  There is also a minor loom related fix to support post call nops included.  I could move that into a separate PR if that would be preferred.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299229](https://bugs.openjdk.org/browse/JDK-8299229): [JVMCI] add support for UseZGC


### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [ce19812e](https://git.openjdk.org/jdk/pull/11996/files/ce19812e0516c8a58cac1a17b7ffbc9c0794a610)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [ce19812e](https://git.openjdk.org/jdk/pull/11996/files/ce19812e0516c8a58cac1a17b7ffbc9c0794a610)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/11996/head:pull/11996` \
`$ git checkout pull/11996`

Update a local copy of the PR: \
`$ git checkout pull/11996` \
`$ git pull https://git.openjdk.org/jdk.git pull/11996/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11996`

View PR using the GUI difftool: \
`$ git pr show -t 11996`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11996.diff">https://git.openjdk.org/jdk/pull/11996.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/11996#issuecomment-1382339123)

fisk reviewed (changes requested):
Please do not remove the patching epoch code for AArch64. It is used by generational ZGC so that it can do self modifying code on AArch64 safely. Generational ZGC is just around the corner. We are currently upstreaming shared runtime code separately so it doesn't need to be merged with the JEP.

tkrodriguez commented:
Ok.  I wondered why it was still around since it seemed pretty obviously dead.  I'll restore that code and expose the extra state needed to support code generation for it.  Are 3 different strategies really required?

fisk commented:
> Ok.  I wondered why it was still around since it seemed pretty obviously dead.  I'll restore that code and expose the extra state needed to support code generation for it.  Are 3 different strategies really required?

We are right at a transition point. We need the STW strategy to not regress for the normal GCs. The conc data + instruction approach, however, is strictly better than the conc data approach. It performs better and allows modifying both data and instructions. But it isn't directly compatible with mainline ZGC, as it kind of wants the colour bits to be enclosed in the low order 32 bits of the thread-local disarm guard value. It is in generational ZGC, but not in mainline ZGC. I suppose it is possible to poke the mainline ZGC code to conform to the new requirements, but as generational ZGC is around the corner, we were sort out waiting for that instead. Presumably something similar would have to be done to Shenandoah as well to migrate to the new better barrier.

tkrodriguez commented:
I've dropped the removal of the epoch stuff and exposed the epoch addr for use by code generation.

fisk commented on src/hotspot/cpu/riscv/gc/shared/barrierSetNMethod_riscv.cpp:
```diff
@@ -50,47 +53,68 @@ static int entry_barrier_offset(nmethod* nm) {
       return -4 * (4 + slow_path_size(nm));
     case NMethodPatchingType::conc_data_patch:
       return -4 * (5 + slow_path_size(nm));
-    case NMethodPatchingType::conc_instruction_and_data_patch:
```
Please re-instate these two lines.

fisk commented on src/hotspot/cpu/x86/gc/shared/barrierSetNMethod_x86.cpp:
```diff
@@ -170,9 +180,16 @@ static const int entry_barrier_offset(nmethod* nm) {
+    barrier_address = nm->code_begin() + nm->jvmci_nmethod_data()->nmethod_entry_patch_offset();
+  } else
+#endif
+  barrier_address = nm->code_begin() + nm->frame_complete_offset() + entry_barrier_offset(nm);
```
This line looks like a ticking bomb. Can we have brackets or something?

fisk commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -561,7 +559,7 @@ nmethod* nmethod::new_nmethod(const methodHandle& method,
   nmethod* nm = NULL;
 #if INCLUDE_JVMCI
-  int jvmci_data_size = !compiler->is_jvmci() ? 0 : JVMCINMethodData::compute_size(nmethod_mirror_name);
+    int jvmci_data_size = compiler->is_jvmci() ? jvmci_data->size() : 0;
```
Weird indentation

fisk commented on src/hotspot/share/jvmci/jvmciCodeInstaller.cpp:
```diff
@@ -727,6 +728,11 @@ JVMCI::CodeInstallResult CodeInstaller::install(JVMCICompiler* compiler,
     }
 
+    if (UseZGC && _nmethod_entry_patch_offset == -1) {
+      // ZGC requires the use of entry barriers for correctness
```
Note that G1 also requires nmethod entry barriers for correctness.

tkrodriguez replied:
We've been happily running without them for a long time which is why we're just now getting around to supporting them.  Are we sitting on a ticking time bomb?  I do want to make them required as of 21 but maybe we could defer that to a future commit once my Graal changes to support the entry barriers has landed?

fisk commented on src/hotspot/share/jvmci/jvmciCodeInstaller.cpp:
```diff
@@ -760,6 +767,20 @@ JVMCI::CodeInstallResult CodeInstaller::install(JVMCICompiler* compiler,
+      if (nm != nullptr) {
+        if (_nmethod_entry_patch_offset != -1) {
+          BarrierSetNMethod* bs_nm = BarrierSet::barrier_set()->barrier_set_nmethod();
+          if (bs_nm == nullptr) {
```
Should never be null. ARM32 was the last to enter the scene but now all platforms have these.

fisk commented on src/hotspot/share/jvmci/jvmciRuntime.hpp:
```diff
@@ -414,7 +457,8 @@ class JVMCIRuntime: public CHeapObj<mtJVMCI> {
                                            FailedSpeculation**       failed_speculations,
                                            char*                     speculations,
-                                           int                       speculations_len);
+                       int                       speculations_len,
```
Weird indentation

fisk commented:
I think synchronization code is needed for accessing metadata from MDOs. And extra care around the extra data section. Stale metadata needs to be filtered, if the metadata is unloading. I don't see any code for that in this patch, so I have to assume it's missing.

tkrodriguez commented:
Sorry I've got too many branches in flight.  Yes I hadn't included the is_loader_alive changes which mirror what's being done in the various CI translate methods that process the same information. I've push those changes.   I'm not sure what locking you're talking about though as I don't see any obvious locking in the reading of this data in the CI.  There is some locking of the extra_data_lock which mainly seems to be guarding against safepoints during this processing.  There are no safepoints between the reading and the check so I'm not sure we need that.

fisk commented:
There are speculative trap entries in the extra data section of the MDOs. Rows with stale metadata must be removed by the GC during class unloading. This used to be done in safepoints only, but with concurrent class unloading this is done while holding the extra_data_lock of the MDO. So naturally, for this to be sound, any reader that traverses these rows must hold the same lock (cf. https://github.com/openjdk/jdk/commit/f08eeac278b0785c0bdff0bddf46b66f65822357).

tkrodriguez commented:
Type profiles are never read from the extra data section by JVMCI and speculative_trap_data_tag isn't exposed at all.  Exception seen flags are the only thing we search for in the extra data section and I'm not even sure why that's done.  See https://github.com/tkrodriguez/jdk/blob/tkr-zgc/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotProfilingInfo.java#L144  and https://github.com/tkrodriguez/jdk/blob/tkr-zgc/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotProfilingInfo.java#L113.  I could adjust the logic to simply disallow reads of Klass* from the extra data section until such a day as we actually need to support it.

fisk commented:
> Type profiles are never read from the extra data section by JVMCI and speculative_trap_data_tag isn't exposed at all. Exception seen flags are the only thing we search for in the extra data section and I'm not even sure why that's done. See https://github.com/tkrodriguez/jdk/blob/tkr-zgc/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotProfilingInfo.java#L144 and https://github.com/tkrodriguez/jdk/blob/tkr-zgc/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotProfilingInfo.java#L113. I could adjust the logic to simply disallow reads of Klass* from the extra data section until such a day as we actually need to support it.

Concurrent class unloading is going to remove rows from the extra data section under the MDO lock. It looks like said lines are iterated over here: https://github.com/tkrodriguez/jdk/blob/tkr-zgc/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotProfilingInfo.java#L163
I don't think the iteration itself is safe, without holding the lock, as rows are concurrently rearranged.

tkrodriguez commented:
I see. MethodData::clean_extra_data repacks the extra data section so it's never safe to operate on that section without holding a lock.  So I don't think this affects the safety of these changes since JVMCI never reads types from that section.  With +UseJVMCICompiler there will never be speculative_trap_data in that section so it will never be shifted.  The only case where that could occur is in Truffle hosted mode where c2 is the JIT but Graal is used for Truffle.  This is presumably a long standing issue, assuming that we actually even use that path.  I'll investigate when we actually read bits data from that section, if at all, and either remove that path or make it safe.

tkrodriguez commented:
I don't understand how this concurrent repacking of the extra data is safe in general.  If an existing record in the extra data section can move at any time then don't operations on that record have to be performed under the lock?  Deoptimization::query_update_method_data may allocate a new record and then use it without caring whether it's in the extra data section or not.  There's some similar JVMCI updates of the exception_seen that have the same problem.  We're probably less exposed since we don't normally have speculative trap entries that would result in repacking.  So it seems to me that every caller of allocate_bci_to_data must be holding the extra_data_lock so it can safely use the data.  Am I missing something?

apangin commented on src/hotspot/share/gc/shared/barrierSetNMethod.cpp:
```diff
@@ -62,11 +65,17 @@ bool BarrierSetNMethod::supports_entry_barrier(nmethod* nm) {
-    return false;
+  if (nm->is_native_method() || nm->is_compiled_by_c2() || nm->is_compiled_by_c1()) {
+    return true;
   }
```
I wonder why `is_native_method` automatically implies `supports_entry_barrier == true`.
Apparently JVMCI is not supposed to compile native methods (without entry barrier), however, this seems too restrictive to me.

Should not we swap the order of the checks? E.g.
```
  if (nm->is_compiled_by_jvmci()) {
    return nm->jvmci_nmethod_data()->has_entry_barrier();
  }

  if (nm->is_native_method() || nm->is_compiled_by_c2() || nm->is_compiled_by_c1()) {
    return true;
  }
```

fisk commented:
> I don't understand how this concurrent repacking of the extra data is safe in general. If an existing record in the extra data section can move at any time then don't operations on that record have to be performed under the lock? Deoptimization::query_update_method_data may allocate a new record and then use it without caring whether it's in the extra data section or not. There's some similar JVMCI updates of the exception_seen that have the same problem. We're probably less exposed since we don't normally have speculative trap entries that would result in repacking. So it seems to me that every caller of allocate_bci_to_data must be holding the extra_data_lock so it can safely use the data. Am I missing something?

You are right, Deoptimization::query_update_method_data does looks like it has a bug. You need to either hold the extra_data_lock when using entries from the extra data section, or force cleaning first, so that existing entries won't move around until the next safepoint poll. The ciMethodData performs cleaning in its replica and keeping all of its type alive, making it safe to use, but extra care needs to be taken when using the MDO directly.

tkrodriguez commented:
/open

tkrodriguez commented:
I've pushed a final round of fixes.  In particular the JVMCI Java code used to read fields in Klass using direct memory access but it must go through the C++ accessors to deal with concurrent unloading.  I also removed the Java side access to the extra data section and replaced it with a new getExceptionSeen accessor which is the only case were we use extra data.  I filed https://bugs.openjdk.org/browse/JDK-8306767 for the method data extra data repacking issue.

dougxc commented on src/jdk.internal.vm.ci/share/classes/jdk/vm/ci/hotspot/CompilerToVM.java:
```diff
@@ -884,23 +884,33 @@ HotSpotConstantPool getConstantPool(MetaspaceObject object) {
+     * only performs the read if the memory location is known to contain a valid Klass*.  If
+     * {@code base} is a {@link HotSpotConstantPool}, {@link HotSpotMethodData}, {@link HotSpotObjectConstantImpl},
+     * or {@link HotSpotResolvedObjectTypeImpl} then the field
+     * corresopnding to {@code displacement} is fetched using the appropriate HotSpot accessor. Any
```
corresopnding -> corresponding

fisk commented:
So do you have AArch64 nmethod entry barriers both for ZGC and the others implemented in Graal? Is it the same instruction sequence as we have?

tkrodriguez commented:
Yes we emit barriers exactly as HotSpot does for both aarch64 and amd64 though we don't support conc_instruction_and_data_patch yet because it's never enabled in the current code.  I refactored NativeNMethodBarrier::check_barrier to verify the barrier at code install time so that we get exceptions if there's a mismatch between HotSpot and Graal, though the verification on the aarch64 side is very weak.  It should probably at least check for the properly placed dmb.

fisk reviewed (approved):
Looks good. Great work, @tkrodriguez.

vnkozlov commented:
Please, run usual testing before integration.

vnkozlov reviewed (approved):
Good.

RealFYang reviewed (changes requested):
Hi, I think you might also want to remove the macro definition of 'COMPRESSED_CLASS_POINTERS_DEPENDS_ON_COMPRESSED_OOPS' in file src/hotspot/cpu/riscv/globalDefinitions_riscv.hpp.

RealFYang commented on src/hotspot/cpu/riscv/gc/shared/barrierSetNMethod_riscv.cpp:
```diff
@@ -57,40 +60,63 @@ static int entry_barrier_offset(nmethod* nm) {
+#if INCLUDE_JVMCI
+    if (nm->is_compiled_by_jvmci()) {
+      _instruction_address = nm->code_begin() + nm->frame_complete_offset();
+      _guard_addr = reinterpret_cast<int*>(nm->consts_begin() + nm->jvmci_nmethod_data()->nmethod_entry_patch_offset());
```
I see 'nm->consts_begin()' is used here to calculate '_guard_addr' for the JVMCI case on riscv. Do you have more details about the design? Thanks.

tkrodriguez replied:
I forgot to update the riscv version since Graal isn't actually fully working there.  It should look just like the aarch64 code in this regard as the same strategy should work there too.

woess commented on src/hotspot/cpu/x86/gc/shared/barrierSetNMethod_x86.cpp:
```diff
@@ -170,9 +180,18 @@ static const int entry_barrier_offset(nmethod* nm) {
+
   NativeNMethodCmpBarrier* barrier = reinterpret_cast<NativeNMethodCmpBarrier*>(barrier_address);
-  debug_only(barrier->verify());
+  barrier->verify();
```
I think this should be reverted to:
`debug_only(barrier->verify());`

tkrodriguez replied:
verify now contains only an assert so the debug_only is unnecessary

woess replied:
I see. Still, `verify()` also contains `err_msg("%s", "");` and that calls `jio_vsnprintf`:
```c++
template <size_t bufsz>
FormatBuffer<bufsz>::FormatBuffer(const char * format, ...) : FormatBufferBase(_buffer) {
  va_list argp;
  va_start(argp, format);
  jio_vsnprintf(_buf, bufsz, format, argp);
  va_end(argp);
}
```
So I think we should probably guard `err_msg("%s", "")` as well if we want to ensure there's no overhead in a product build.

tkrodriguez commented:
I think I've addressed all the review comments and the mach5 testing and Graal gating on these changes are all clean.

stefank commented on src/hotspot/share/jvmci/jvmciCompilerToVM.hpp:
```diff
@@ -25,6 +25,7 @@
 #define SHARE_JVMCI_JVMCICOMPILERTOVM_HPP
 
 #include "gc/shared/cardTable.hpp"
+#include "gc/shared/barrierSetAssembler.hpp"
```
Sort includes.

stefank commented on src/hotspot/share/jvmci/jvmciCompilerToVMInit.cpp:
```diff
@@ -26,10 +26,14 @@
 #include "gc/shared/cardTable.hpp"
 #include "gc/shared/collectedHeap.hpp"
 #include "gc/shared/gc_globals.hpp"
 #include "gc/shared/tlab_globals.hpp"
+#include "gc/shared/barrierSetNMethod.hpp"
```
Sort includes.

stefank commented on src/hotspot/share/jvmci/jvmciCompilerToVMInit.cpp:
```diff
@@ -26,10 +26,14 @@
 #include "gc/shared/gc_globals.hpp"
 #include "gc/shared/tlab_globals.hpp"
+#include "gc/shared/barrierSetNMethod.hpp"
+#include "gc/z/zThreadLocalData.hpp"
+#include "gc/z/zBarrierSetRuntime.hpp"
```
Should this be guarded with an INCLUDE_ZGC check? This also needs to be sorted.

tkrodriguez replied:
Guarded the includes and the code which uses them.

stefank commented on src/hotspot/share/jvmci/jvmciEnv.hpp:
```diff
@@ -27,6 +27,7 @@
 
 #include "classfile/javaClasses.hpp"
 #include "jvmci/jvmciJavaClasses.hpp"
+#include "oops/klass.inline.hpp"
```
Don't include .inline.hpp file in .hpp files.

stefank commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -866,6 +858,7 @@ nmethod::nmethod(
 #if INCLUDE_JVMCI
     _speculations_offset     = _nul_chk_table_offset + align_up(nul_chk_table->size_in_bytes(), oopSize);
     _jvmci_data_offset       = _speculations_offset  + align_up(speculations_len, oopSize);
+    int jvmci_data_size = compiler->is_jvmci() ? jvmci_data->size() : 0;
```
Indentation seems off.

stefank commented on src/hotspot/share/gc/shared/barrierSetNMethod.hpp:
```diff
@@ -28,6 +28,9 @@
 #include "memory/allocation.hpp"
 #include "utilities/globalDefinitions.hpp"
 #include "utilities/sizes.hpp"
+#if INCLUDE_JVMCI
+#include "utilities/formatBuffer.hpp"
+#endif
```
Given that this doesn't include any JVMCI files I think we can skip the INCLUDE_JVMCI guard (and sort in the added include). Or maybe better, forward declare FormatBuffer?

stefank commented on src/hotspot/share/jvmci/jvmciCodeInstaller.cpp:
```diff
@@ -760,6 +770,17 @@ JVMCI::CodeInstallResult CodeInstaller::install(JVMCICompiler* compiler,
       }
+
+      if (nm != nullptr) {
+        if (_nmethod_entry_patch_offset != -1) {
+          FormatBuffer<> msg("%s", "");
+          BarrierSetNMethod* bs_nm = BarrierSet::barrier_set()->barrier_set_nmethod();
+
+          if (!bs_nm->verify_barrier(nm, msg)) {
+            JVMCI_THROW_MSG_(IllegalArgumentException, err_msg("nmethod entry barrier is malformed: %s", msg.buffer()), JVMCI::ok);
+          }
+        }
+      }
```
We have this typedef:
```
typedef FormatBuffer<> err_msg;

```
So, it looks weird to have code that mixes FormatBuffer<> and err_msg.

tkrodriguez replied:
I converted all bare usages of FormatBuffer<> to err_msg which seems to be the general usage.

stefank reviewed (commented):
This adds support for the non-generational ZGC. We are currently in the process of reviewing Generational ZGC. Have @fisk talked to you about that? It's unclear to me if Generational ZGC will immediately break these changes. Erik could you shed some light w.r.t that?

I took the opportunity to look at the patch to see if we would get any major merge conflicts. While looking through the patch I found a few nits that I think would be nice to get fixed.

tkrodriguez commented:
I spoke some with Erik about generational support and it will require a fair amount of work for Graal support.  In particular the emitted code sequences have gotten more complex and we need new MarkIds for the new relocs.  I can't speak to whether it breaks these changes but we can always address them in a later issue if it does.  We can't reliably track the very latest master anyway.

stefank commented:
OK. Thanks for the info.

tkrodriguez commented:
/integrate
